src/boot/efi/util.c

```diff
@@ -769,6 +770,111 @@ EFI_STATUS make_file_device_path(EFI_HANDLE device, const char16_t *file, EFI_DE
+
+/* detect direct boot */
+bool is_direct_boot(EFI_HANDLE device) {
+        EFI_STATUS err;
+        VENDOR_DEVICE_PATH *dp;
+
+        err = BS->HandleProtocol(device, &DevicePathProtocol, (void **) &dp);
+        if (err != EFI_SUCCESS)
+                return false;
+
+        /* 'qemu -kernel systemd-bootx64.efi' */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_VENDOR_DP &&
+            memcmp(&dp->Guid, &(EFI_GUID)QEMU_KERNEL_LOADER_FS_MEDIA_GUID, sizeof(EFI_GUID)) == 0)
+                return true;
+
+        /* loaded from firmware volume (sd-boot added to ovmf) */
+        if (dp->Header.Type == MEDIA_DEVICE_PATH &&
+            dp->Header.SubType == MEDIA_PIWG_FW_VOL_DP)
+                return true;
+
+        return false;
+}
+
+/* return the length of the device path in bytes, excluding the end tag */
+static size_t device_path_length(const EFI_DEVICE_PATH *dp) {
+        assert(dp);
+
+        const EFI_DEVICE_PATH *end_node = dp;
+        while (!IsDevicePathEnd(end_node))
+                end_node = NextDevicePathNode(end_node);
+        return (uint8_t*)end_node - (uint8_t*)dp;
+}
+
+/* try find ESP when not loaded from ESP */
+EFI_STATUS qemu_open(EFI_HANDLE *ret_qemu_dev, EFI_FILE **ret_qemu_dir) {
+        _cleanup_free_ EFI_HANDLE *handles = NULL;
+        size_t n_handles;
+        EFI_STATUS err;
+
+        assert(ret_qemu_dev);
+        assert(ret_qemu_dir);
+
+        /* ovmf doesn't connect devices for direct kernel boot */
+        reconnect();
+
+        /* find all file system handles */
+        err = BS->LocateHandleBuffer(ByProtocol, &FileSystemProtocol, NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return err;
+
+        for (size_t order = 0;; order++) {
+                _cleanup_free_ EFI_DEVICE_PATH *dp = NULL;
+                char16_t order_str[20];
+                size_t dp_len = 0;
+
+                SPrint(order_str, sizeof(order_str), u"QemuBootOrder%04x", order);
+                err = efivar_get_raw(&(EFI_GUID)QEMU_BOOT_ORDER_GUID, order_str, (char**)&dp, NULL);
+                if (err == EFI_SUCCESS)
+                        dp_len = device_path_length(dp);
+
+                for (size_t i = 0; i < n_handles; i++) {
+                        _cleanup_(file_closep) EFI_FILE *root_dir = NULL, *efi_dir = NULL;
+                        EFI_DEVICE_PATH *fs;
+                        size_t fs_len;
+
+                        err = BS->HandleProtocol(handles[i], &DevicePathProtocol, (void **) &fs);
+                        if (err != EFI_SUCCESS)
+                                return err;
+                        fs_len = device_path_length(fs);
+
+                        /* check against QemuBootOrderNNNN (if set) */
+                        if (dp_len > fs_len)
+                                continue;
+                        if (dp_len > 0 && memcmp(dp, fs, dp_len) != 0)
+                                continue;
```

Comment:
this is not performance sensitive, so why not add a nicely abstracted device_path_equal() helper? even if this means calling device_path_length() a bit more often than like this, it shouldn't really matter, and is much easier to follow...

(or is there DevicePathEqual() defined already in some api?)

Reply:
We don't need equal, we need substring, fs usually is longer, i.e.

fs = root / controller / disk / partition / ,,,
dp = root / controller / disk

Reply:
then maybe add `device_path_startswith()`?

That would nicely mimic the `startswith()` and `path_startswith()` calls that we have in our userspace helper libraries